src/Core/Checkout/Promotion/DataAbstractionLayer/PromotionRedemptionUpdater.php

```diff
@@ -329,4 +354,26 @@ private function getAllCustomerCounts(array $promotionIds): array
 
         return $allCustomerCounts;
     }
+
+    /**
+     * @param list<string> $promotionIds
+     *
+     * @return array<string, int>
+     */
+    private function fetchPromotionCounts(array $promotionIds): array
+    {
+        $criteria = new Criteria();
+        $criteria->addFilter(new EqualsFilter('versionId', Defaults::LIVE_VERSION));
+        $criteria->addFilter(new EqualsFilter('type', PromotionProcessor::LINE_ITEM_TYPE));
+        $criteria->addFilter(new EqualsAnyFilter('promotionId', $promotionIds));
+
+        $lineItems = $this->lineItemRepository->search($criteria, Context::createDefaultContext())->getEntities();
```

Comment:
it seems highly inefficient to do this query via the DAL with hydrating objects here. I would vote for a simple SQL query with `COUNT()` and `GROUP BY promotion_id`.